jetbrains: fix remote dev server for IDEs

tymscar opened the pull request:
###### Description of changes

Fixes #153335 for most of the Jetbrains tools

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

tymscar committed:
jetbrains: fix remote dev server for IDEs

Adds the necessary patch to fix the remote dev server in the Jetbrains IDEs ( see #153335 ).

tymscar commented:
Have tested and it works on:
- clion
- goland 
- idea-ultimate
- phpstorm 
- pycharm-professional
- ruby-mine
- webstorm 

To run it after install you need to use the binary that comes with each ide, for example for webstorm its `webstorm-remote-dev-server run <path to project>`. Then it should appear on the client in the SSH tab inside of Gateway

tymscar committed:
maintainer: added myself (tymscar) as a maintainer to some Jetbrain IDEs

GenericNerdyUsername commented:
If youre applying a patch to every ide, its probably a better idea to put it in `linux.nix` and/or `darwin.nix`

tymscar commented:
Hey! I was thinking of adding it to linux.nix but the issue is that some IDE's dont have remote dev support such as the community editions of intellij and pycharm.

> why is this a problem?

I am not entirely sure why, but if `REMOTE_DEV_SERVER_USE_SELF_CONTAINED_LIBS` is 0 when the IDE launches, it will not work. It was my preffered method, because doing it like that would mean no patch was needed at all, but whenever I would set it to 0 it wouldn't even boot the server anymore.

GenericNerdyUsername commented:
Could you add a test to see if the plugins directory contains the right folder, and only patch then?

tymscar commented:
Great idea @GenericNerdyUsername !

I have added a new commit where I moved all the logic for the patching in `linux.nix` as you've initially suggested and made it so I check for the existance of the right folder so now all packages work with it. This made `default.nix` way cleaner.

I have tried every single package exposed by this and they all work with the remote extension except dataspell and datagrip which are more different, so I am ignoring those. Those still however work normally, I have tested that!

![image](https://github.com/NixOS/nixpkgs/assets/3742502/700c19e9-3319-406a-87a8-02c6d7dc458b)

GenericNerdyUsername commented on pkgs/applications/editors/jetbrains/linux.nix:
```diff
@@ -119,6 +123,18 @@ with stdenv; lib.makeOverridable mkDerivation (rec {
+      fi
+    fi
+  '';
+
```
Would it be possible to remove this if `$out/bin/clion` (the old place for the wrapper) was a symlink to `$out/clion/bin/clion`(the new place for the wrapper)?

tymscar replied:
I'm not sure I understand the question. You mean removing the whole postInstall in that case?
I think it's needed not only for the location, but also for the env variables, but I might be wrong.

GenericNerdyUsername replied:
No, I mean instead of wrapping `$out/clion/bin/clion` and putting the result in `$out/bin/clion`, move `$out/clion/bin/clion` to  `$out/clion/bin/.clion_wrapped`, put the wrapper at  `$out/clion/bin/clion` (so the remote ide launch script uses it) and make `$out/bin/clion` a symlink to `$out/clion/bin/clion`

tymscar replied:
Well the thing I would be wrapping is not `clion` in this case, it would be `$out/clion/bin/remote-dev-server.sh`.
There are 2 executables that get installed `clion` and `clion-remote-dev-server`. 

`Clion` is the normal IDE in this case and is a binary, and `clion-remote-dev-server` is a script that sets up a bunch of variables and at the end calls `remote-dev-server.sh` which is what this is, wrapped.

If I just symlink clion to it, I lose the normal IDE, and I only get the remote extension functionality.

I might misunderstand your comment.

GenericNerdyUsername replied:
The `remote-dev-server.sh` script calls the normal `$out/clion/bin/clion.sh` script, just with some extra args. I doubt any of the bash scripts need any of the programs added by the wrapper, so if we wrap `$out/clion/bin/clion.sh`, everything gets the extra programs in one place

tymscar replied:
Are you saying that instead of having then two executables, the user will end up having just one, in this case Clion, that will be the wrapped script, and because both call the same initial script, if launched without the “run <path to project>” it could just run the IDE like `clion` runs it now, and if provided with those parameters it will run it like `clion-remote-dev-server` does it now?

GenericNerdyUsername replied:
there will be two user-facing executables:
- `$out/bin/clion`: a symlink to `$out/clion/bin/clion.sh`, which is a wrapper script generated by us to add some stuff to the path, then call `$out/clion/bin/.clion_wrapped`
- `$out/bin/clion-remote-dev-server`: a symlink to `$out/clion/bin/remote-dev-server.sh`, which calls `$out/clion/plugins/remote-dev-server/bin/launcher.sh` (with some args), which calls  `$out/clion/bin/clion.sh` (with some args), which calls `$out/clion/bin/.clion_wrapped`

tymscar replied:
I fully understand now. I didn't realise you were talking about the previous wrapper in the installPhase.

I have added a new commit removing the second wrapper, and I've tested it and it all still works great.

The reason why I created a wrapping script for the remote dev server as well is because the original script contains `$0` references that broke when I tried with a simple symlink, but now there is no more duplication about the makeWrapper variables!

GenericNerdyUsername commented on pkgs/applications/editors/jetbrains/default.nix:
```diff
@@ -17,6 +17,13 @@
+, git
+, libsecret
+, e2fsprogs
+, libnotify
```
Is this still needed, now that the other wrapper has been removed?

tymscar replied:
Great catch! I have removed them now, and tested just to be sure!

GenericNerdyUsername commented on pkgs/applications/editors/jetbrains/linux.nix:
```diff
@@ -99,7 +103,9 @@ with stdenv; lib.makeOverridable mkDerivation (rec {
-    makeWrapper "$out/$pname/bin/${loName}.sh" "$out/bin/${pname}" \
+    mv $out/$pname/bin/${loName}.sh $out/$pname/bin/${loName}_wrapped.sh
+
+    makeWrapper "$out/$pname/bin/${loName}_wrapped.sh" "$out/$pname/bin/${loName}.sh" \
```
I think this can be replaced with [`wrapProgram`](https://nixos.org/manual/nixpkgs/unstable/#fun-wrapProgram)

tymscar replied:
I didn't know about wrapProgram. That's very useful! I learned a lot during this PR! Thank you!

GenericNerdyUsername reviewed (changes requested):
After this, I think the only things needed before merge are squashing/splitting the commits into two, one for adding yourself as a maintainer and one for everything else, then it'll be ready for merge (in my view)

tymscar commented:
> After this, I think the only things needed before merge are squashing/splitting the commits into two, one for adding yourself as a maintainer and one for everything else, then it'll be ready for merge (in my view)

What is the usual process in nixpkgs before merging? Do I need every owner to approve?

Vinetos commented:
Works well here :D Thanks !

Can't wait to be merged :+1:

SuperSandro2000 commented on pkgs/applications/editors/jetbrains/linux.nix:
```diff
@@ -114,11 +118,14 @@ with stdenv; lib.makeOverridable mkDerivation (rec {
 
+    ln -s "$out/$pname/bin/${loName}.sh" $out/bin/$pname
+    echo -e '#!/usr/bin/env bash\n'"$out/$pname/bin/remote-dev-server.sh"' "$@"' > $out/$pname/bin/remote-dev-server-wrapped.sh
+    chmod +x $out/$pname/bin/remote-dev-server-wrapped.sh
```
Do we actually need this twice?

tymscar replied:
Yes because the script references `$0` and any link to it would break when ran from outside that folder.

kirinnee commented:
Works on linux ubuntu for clion, goland, webstorm! can't wait for this to be merged

pbek commented:
Does this maybe cause: https://github.com/NixOS/nixpkgs/pull/242493#issuecomment-1643419680?
